File: weasel.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <time.h>
#include <ctype.h>
#include <getopt.h>

#define STRSIZE                 258
#define DEFAULT_POPULATION_SIZE 100
#define DEFAULT_START_LEN        10
#define DEFAULT_POINT_PROB        1
#define DEFAULT_DUPLICATE_PROB    1
#define DEFAULT_DELETE_PROB       1
#define DEFAULT_TRANSPOSE_PROB    1
#define DEFAULT_TARGET_STRING   "methinks it is like a weasel"
#define DEFAULT_ALPHABET        " abcdefghijklmnopqrstuvwxyz" \
                                "ABCDEFGHIJKLMNOPQRSTUVWXYZ" \
			        "0123456789.,?!\\'"

/**
 * Possible mutation events.  
 */
enum MutationEvent {
    NO_CHANGE,          // a  -> a 
    FIRST_EVENT = NO_CHANGE,
    POINT_CHANGE,       // a  -> b
    DUPLICATE,          // a  -> aa
    DELETE,             // ab -> b
    TRANSPOSE,          // ab -> ba
    LAST_EVENT = TRANSPOSE
};

/**
 * Return a random value in the range [min,max]
 */
int randInRange( int min, int max )
{
  double scale = 1.0 / RAND_MAX;
  double range = max - min + 1;
  return min + (int) (scale * range * rand() );
}

/**
 * Write a random sequence of len characters to target
 */
void randomString( char *target, size_t len, const char *alphabet )
{
  size_t asize = strlen( alphabet );
  for ( size_t i = 0; i < len-1; i++ )
    *target++ = alphabet[randInRange(0,asize-1)];
  *target = 0;
}

/**
 * Compute the score of the candidate string
 */
int score( const char *candidate, const char *target )
{
  int finalScore = 0;
  while ( *candidate && *target )
  {
    int tmp = *candidate++ - *target++;
    if ( tmp < 0 )
      tmp = -tmp;
    finalScore += tmp;
  }
  while ( *candidate )
    finalScore += *candidate++;
  while( *target )
    finalScore += *target++;

  return finalScore;
}

/**
 * Set up the mutation event probability table
 */
void setupProbTable( int *probTable, int point, int duplicate, int delete, 
                     int transpose )
{
  int probSum = point + duplicate + delete + transpose;
  probTable[NO_CHANGE]    = 100 - probSum;
  probTable[POINT_CHANGE] = probTable[NO_CHANGE] + point;
  probTable[DUPLICATE]    = probTable[POINT_CHANGE] + duplicate;
  probTable[DELETE]       = probTable[DUPLICATE] + delete;
  probTable[TRANSPOSE]    = probTable[DELETE] + transpose;
}

/**
 * Get the next mutation event, based on that event's probability
 */
enum MutationEvent getMutationEvent( int *probTable )
{
  int roll = randInRange(1,100);
  for ( enum MutationEvent e = FIRST_EVENT; e <= LAST_EVENT; e++ )
  {
    if ( roll < probTable[e] )
      return e;
  }
  return NO_CHANGE;
}
  
/**
 * Make a mutated copy of the input string; returns the mutation rate for 
 * that string
 */
double mutate( const char *src, char *dst, int *probTable, 
                const char *alphabet )
{
  size_t alen = strlen( alphabet );
  int chars = 0;
  int mutCount = 0;

  while ( *src )
  {
    chars++;
    switch ( getMutationEvent( probTable ) )
    {
      case NO_CHANGE: 
        *dst++ = *src++; 
        break;

      case POINT_CHANGE: 
        *dst++ = alphabet[randInRange(0, alen-1)]; 
        src++;
        mutCount++; 
        break;

      case DUPLICATE: 
        *dst++ = *src; 
        *dst++ = *src++;
        mutCount++; 
        break;

      case DELETE: 
        src++; 
        mutCount++;
        break;

      case TRANSPOSE:
        if ( *(src + 1) )
        {
          *dst++ = *(src + 1);
          *dst++ = *src;
          src += 2;
          mutCount++;
        }
        break;
    }
  }
  *dst = 0;
  return (double) mutCount / (double) chars;
}

/**
 * Display usage statement
 */
void usage( const char *progName, const char *error )
{
  const char *msg =
    "USAGE: %s -h or \n"
    "       %s [mutation parameters] [search parameters] [logging parameters]\n"
    "\n"
    "mutation parameters:\n"
    "  -p n - point change probability\n"
    "  -d n - duplication probability\n"
    "  -x n - deletion probability\n"
    "  -t n - transposition probability\n"
    "\"n\" is an integer between 0 and 100 (default is 1)\n"
    "mutation probabilities must sum to less than or equal to 100\n"
    "\n"
    "search parameters:\n"
    "  -S n   - population size; n must be an integer greater than 10 (default is %d)\n"
    "  -T str - target string (default is \"%s\")\n"
    "  -A str - alphabet for new strings (default is \"%s\")\n"
    "  -l n   - initial string size (default is %d)\n"
    "\n"
    "logging parameters\n"
    "  -L logfile - writes detailed logging information to logfile (default is no detailed logging)\n";

  if ( error != NULL && strcmp( error, "" ) != 0 )
    fprintf( stderr, "ERROR: %s\n", error );
    
  fprintf( stderr, msg, progName, progName, DEFAULT_POPULATION_SIZE, DEFAULT_TARGET_STRING, DEFAULT_ALPHABET, DEFAULT_START_LEN );
  exit( 0 );
}

/**
 * Process command line parameters
 */
void getParameters( int argc, char **argv, int *point, int *duplicate, 
                     int *delete, int *transpose, size_t *popSize, 
                     char *target, char *alphabet, int *len, 
                     int *log, char *logFileName )
{
  int opt;

  while ( (opt = getopt(argc, argv, "hp:d:x:t:S:T:A:l:L:")) != -1 )
  {
    switch ( opt )
    {
      char *chk;
      case 'h':
        usage( argv[0], NULL );
        exit(0);
        break;

      case 'p':
        *point = (int) strtol( optarg, &chk, 10 );
        if ( ( !isspace( *chk ) && *chk != 0 ) || ( *point < 0 || *point > 100 ))
          usage( argv[0], "invalid point change argument" );
        break;

      case 'd':
        *duplicate = (int) strtol( optarg, &chk, 10 );
        if ( ( !isspace( *chk ) && *chk != 0 ) || ( *duplicate < 0 || *duplicate > 100 ))
          usage( argv[0], "invalid duplicate argument" );
        break;

      case 'x':
        *delete = (int) strtol( optarg, &chk, 10 );
        if ( ( !isspace( *chk ) && *chk != 0 ) || ( *delete < 0 || *delete > 100 ))
          usage( argv[0], "invalid delete argument" );
        break;

      case 't':
        *transpose = (int) strtol( optarg, &chk, 10 );
        if ( ( !isspace( *chk ) && *chk != 0 ) || ( *transpose < 0 || *transpose > 100 ))
          usage( argv[0], "invalid transpose argument" );
        break;

      case 'S':
        *popSize = (size_t) strtoul( optarg, &chk, 10 );
        if ( ( !isspace( *chk ) && *chk != 0 ) || *popSize < 10 )
          usage( argv[0], "invalid population size" );
        break;

      case 'T':
        strcpy( target, optarg );
        break;

      case 'A':
        strcpy( alphabet, optarg );
        break;

      case 'l':
        *len = (int) strtol( optarg, &chk, 10 );
        if ( !isspace( *chk ) && *chk != 0 )
          usage( argv[0], "invalid string length" );
        break;

      case 'L':
        *log = 1;
        strcpy( logFileName, optarg );
        break;
    }
  }

  if ( *point + *duplicate +*delete + *transpose > 100 )
  {
    usage( argv[0], "The sum of all mutation probabilities must be <= 100" );
  }
}

void getParametersString( const char *str, int *point, int *duplicate, int *delete, int *transpose,
                          size_t *popSize, char *target, char *alphabet, int *len )
{
  fprintf( stderr, "checking for query string\n" );

  if ( str )
  {
    char *tmp = malloc( strlen( str ) + 1 );
    strcpy( tmp, str );

    char *keyvals[8] = {NULL};
    size_t i = 0;

    keyvals[0] = strtok( tmp, "&" );
    while ( i < 8 && keyvals[i] )
      keyvals[++i] = strtok( NULL, "&" );

    for ( i = 0; i < 8 && keyvals[i]; i++ )
    {
      char *key = strtok( keyvals[i], "=" );
      char *value = strtok( NULL, "=" );

      fprintf( stderr, "key = %s, value = %s\n", key, value );

      if ( !strcmp( key, "target" ) )
      {
        char *nxt = value;
        while ( (nxt = strchr( nxt, '+' ) ) )
          *nxt = ' ';
        strcpy( target, value );
      }
      else if ( !strcmp( key, "alphabet" ) )
      {
        char *nxt = value;
        while ( ( nxt = strchr( nxt, '+' ) ) )
          *nxt = ' ';
        strcpy( alphabet, value );
      }
      else if ( !strcmp( key, "point" ) )
      {
        *point = (int) strtol( value, NULL, 0 );
      }
      else if ( !strcmp( key, "duplicate" ) )
      {
        *duplicate = (int) strtol( value, NULL, 0 );
      }
      else if ( !strcmp( key, "delete" ) )
      {
        *delete = (int) strtol( value, NULL, 0 );
      }      
      else if ( !strcmp( key, "transpose" ) )
      {
        *transpose = (int) strtol( value, NULL, 0 );
      }
      else if ( !strcmp( key, "popSize" ) )
      {
        *popSize = (size_t) strtoul( value, NULL, 0 );
      }
      else if ( !strcmp( key, "startLen" ) )
      {
        *len = (int) strtol( value, NULL, 0 );
      }
    }
    free( tmp );
  }
  else
  {
    fprintf( stderr, "No inputs\n" );
  }
}

/**
 * Data type for each weasel
 */
struct weasel {
  char str[STRSIZE];
  int score;
  unsigned long generation;
  double mutRate;
};

/**
 * Compare two weasels; called by qsort
 */
int cmpWeasel( const void *l, const void *r )
{
  const struct weasel *lw = l;
  const struct weasel *rw = r;

  return lw->score - rw->score;
}

/**
 * Write progress to a log file on disk.
 */
void progress( struct weasel *population, size_t popSize, FILE *log, 
                int width )
{
  size_t breedPopSize = popSize / 10;
  size_t column = 0;
 
  fprintf( log, "\nGeneration: %lu\n\n", population[breedPopSize].generation);
  fprintf( log, "Breeding population:\n");
  column = 0;
  for ( size_t i = 0; i < breedPopSize; i++ )
  {
    fprintf( log, "%4zu: %7.4f %-*.*s%s", i, population[i].mutRate, width, 
      width, population[i].str, ++column % 3 == 0 ? "\n" : "    " );
  }
  fprintf( log, "\n\nProgeny:\n" );
  column = 0;
  for ( size_t i = breedPopSize; i < popSize; i++  )
  {
    fprintf( log, "%4zu: %7.4f %-*.*s%s", i, population[i].mutRate, width, 
      width, population[i].str, ++column % 3 == 0 ? "\n" : "    " );
  }
  fputc( '\n', log );
}
  
int main( int argc, char **argv )
{
  srand( time( NULL ) );
  int point = DEFAULT_POINT_PROB;
  int duplicate = DEFAULT_DUPLICATE_PROB;
  int delete = DEFAULT_DELETE_PROB;
  int transpose = DEFAULT_TRANSPOSE_PROB;
  int probTable[] = { 96, 97, 98, 99, 100 };
  char alphabet[STRSIZE] = DEFAULT_ALPHABET;
  char target[STRSIZE] = DEFAULT_TARGET_STRING;
  struct weasel fittest = {"", 0, 0};
  size_t popSize = DEFAULT_POPULATION_SIZE;
  unsigned long generation = 0UL;
  int startLen = 5;
  int writeLog = 0;
  FILE *log = NULL;
  char logName[512] = "weasel.log";

  if ( getenv( "GATEWAY_INTERFACE" ) != NULL && strcmp( getenv( "GATEWAY_INTERFACE" ), "" ) != 0 )
  {
    /**
     * We were invoked by a Web client through the Common Gateway Interface, so we need to
     * create an HTML form for input.
     */
    printf( "Content-type: text/html\r\n\r\n" );
    printf( "<!DOCTYPE HTML><html><head><title>Pop the Weasel!!!</title>" );
    printf( "<link rel=\"stylesheet\" href=\"../weasel.css\"></head>" );
    printf( "<body><p>This is my particular spin on Dawkins' WEASEL program as descibed in <em>The Blind Watchmaker</em></p>" );
    printf( "<form action=\"./weasel\" method=\"post\">" );
    printf( "<label for=\"target\">Target string:</label><input type=\"text\" name=\"target\" value=\"%s\"><br>", DEFAULT_TARGET_STRING );
    printf( "<label for=\"alphabet\">Alphabet:</label><input type=\"text\" name=\"alphabet\" value=\"%s\"><br>", DEFAULT_ALPHABET );
    printf( "<label for=\"point\">Point change probability:</label><input type=\"text\" name=\"point\" value=\"%d\"><br>", DEFAULT_POINT_PROB );
    printf( "<label for=\"duplicate\">Duplication probability</label><input type=\"text\" name=\"duplicate\" value=\"%d\"><br>", DEFAULT_DUPLICATE_PROB );
    printf( "<label for=\"delete\">Deletion probability</label><input type=\"text\" name=\"delete\" value=\"%d\"><br>", DEFAULT_DELETE_PROB );
    printf( "<label for=\"transpose\">Transposition probability</label><input type=\"text\" name=\"transpose\" value=\"%d\"><br>", DEFAULT_TRANSPOSE_PROB );
    printf( "<label for=\"popSize\">Population size</label><input type=\"text\" name=\"popSize\" value=\"%d\"><br>", DEFAULT_POPULATION_SIZE );
    printf( "<label for=\"startLen\">Start string length</label><input type=\"text\" name=\"startLen\" value=\"%d\"><br>", DEFAULT_START_LEN );
    printf( "<input type=\"submit\" name=\"submitbutton\" id=\"submitbutton\" value=\"Pop That Weasel!\">" );
    printf( "</form><br><br>" );
  }

  char *method = getenv( "REQUEST_METHOD" );

  if ( method && strcmp( method, "GET" ) == 0 )
  {
    /**
     * We were invoked via an HTTP GET method, so we pull the input parameters
     * from the QUERY_STRING environment variable.
     */
    fprintf( stderr, "calling getParametersString with QUERY_STRING\n" );

    getParametersString( getenv( "QUERY_STRING" ), &point, &duplicate, &delete, &transpose, 
                        &popSize, target, alphabet, &startLen );
  }
  else if ( method && strcmp( method, "POST" ) == 0 )
  {
    /**
     * We were invoked via an HTTP POST method, so we pull the input parameters
     * from the standard input streaqm.
     */
    fprintf( stderr, "calling getParametersString with POST input\n" );
    char input[2048];
    if ( fgets( input, sizeof input, stdin ) )
    {
      getParametersString( input, &point, &duplicate, &delete, &transpose,
                           &popSize, target, alphabet, &startLen );
    }
  }
  else
  {    
    /**
     * We were invoked from the command line in a terminal session
     */
    getParameters( argc, argv, &point, &duplicate, &delete, 
                   &transpose, &popSize, target, alphabet, &startLen,
                   &writeLog, logName );
  }

  /**
   * Make sure the target doesn't contain any characters that are not
   * in the source character set.
   */
  if ( strspn( target, alphabet ) < strlen( target ) )
  {
    usage( argv[0], "Target string contains characters not in the alphabet!" );
    exit( -1 );
  }

  /**
   * Initialize our probability table.
   */
  setupProbTable( probTable, point, duplicate, delete, transpose );

  /**
   * Allocate memory for our Weasel population.
   */
  struct weasel *population = malloc( sizeof *population * popSize );
  if ( population )
  {
    /**
     * Set up the progress log file
     */
    if ( writeLog )
    {
      log = fopen( logName, "w" );
      if ( !log )
      {
        fprintf( stderr, "Could not open log file %s; will not log progress",
                 logName );
        writeLog = 0;
      }
    }

    /**
     * Generate the initial population of random strings and score them
     */
    for ( size_t i = 0; i < popSize; i++ )
    {
      randomString( population[i].str, startLen, alphabet );
      population[i].score = score( population[i].str, target );
      population[i].generation = generation;
      population[i].mutRate = 0.0;
    }

    /**
     * Sort the population by score
     */
    qsort( population, popSize, sizeof *population, cmpWeasel );

    if ( getenv( "GATEWAY_INTERFACE" ) && strcmp( getenv( "GATEWAY_INTERFACE" ), "" ) != 0 )
    {
      printf( "<table><tr><th>Generation</th><th>Score</th><th>Value</th></tr>" );
    }
    else
    {
      printf( "%15s%15s%8s%s\n", "Generation", "Score", " ", "Value" );
      printf( "%15s%15s%8s%s\n", "----------", "-----", " ", "-----" );
    }
    /**
     * As long as we don't have a match (minimum score is not 0), take the top
     * 10% of the population and make mutated copies of them, writing over the
     * bottom 90% of the population.  Display the individual with the lowest
     * score (will only be displayed once).
     */
    while ( population[0].score > 0 )
    {
      generation++;

      if ( strcmp( fittest.str, population[0].str ))
      {
        if ( getenv( "GATEWAY_INTERFACE" ) && strcmp( getenv( "GATEWAY_INTERFACE" ), "" ) != 0 )
        {
          printf( "<tr><td>%lu</td><td>%d</td><td>%s</td></tr>", population[0].generation,
            population[0].score, population[0].str );
        }
        else
        {
          printf( "%15lu%15d%8s%s\n", population[0].generation, 
                                      population[0].score, " ", 
                                      population[0].str );
        }
        fittest = population[0];
      }

      size_t breedPopSize = popSize / 10;
      for ( size_t i = breedPopSize; i < popSize; i++ )
      {
        population[i].mutRate = mutate( population[i%breedPopSize].str, 
                                        population[i].str, probTable, 
                                        alphabet );
        population[i].score = score( population[i].str, target );
        population[i].generation = generation;
      }
      if ( writeLog )
        progress( population, popSize, log, strlen( target ) );

      qsort( population, popSize, sizeof *population, cmpWeasel );
    }
    if ( getenv( "GATEWAY_INTERFACE" ) && strcmp( getenv( "GATEWAY_INTERFACE" ), "" ) != 0 )
    {
      printf( "<tr><td>%lu</td><td>%d</td><td>%s</td></tr></table>", population[0].generation,
        population[0].score, population[0].str );
    }
    else
    {
      printf( "%15lu%15d%8s%s\n", population[0].generation, 
                                  population[0].score, " ", 
                                  population[0].str );
    }
    free( population );    

    if ( getenv( "GATEWAY_INTERFACE" ) && strcmp( getenv( "GATEWAY_INTERFACE" ), "" ) != 0 )
    {
      printf( "</body></html>" );
    }
  }
  if ( writeLog )
    fclose( log );

  return 0;
}

File: weasel.js
/**
 * Generate a random string of the given length from the
 * source alphabet. 
 *
 * Inputs:
 *     length   - output string length
 *     alphabet - character set from which to generate the string
 *
 * Outputs:
 *     none
 *
 * Returns: Randomly generated string of the specified length  
 */
function randomString( length, alphabet )
{
  var arr = [];
  var i;

  for ( i = 0; i < length; i++ )
  {
    arr[i] = alphabet.charAt( Math.random() * alphabet.length );
  }
  return arr.join("");
}

/**
 * Mutate the input string based on a probability table.
 *
 * Inputs:
 *    candidate   - source string to mutate
 *    prob_table  - list of probabilities for each mutation type
 *    alphabet    - character set from which to choose point change characters
 *
 * Outputs:
 *    none
 *
 * Returns: a mutated copy of the candidate string
 */
function mutate( candidate, prob_table, alphabet )
{
  var arr = [];
  var r = 0;
  var w = 0;
  var c;
  
  for ( r = 0; r < candidate.length; r++)
  {
    var event = Math.random() * 100;
    if ( event < prob_table[0] ) // no change
    {
      arr[w++] = candidate[r];
    }
    else if ( event >= prob_table[0] && event < prob_table[1] ) // point change
    {
      arr[w++] = alphabet.charAt( Math.random() * alphabet.length );
    }
    else if ( event >= prob_table[1] && event < prob_table[2] ) // delete
    {
    }
    else if ( event >= prob_table[2] && event < prob_table[3] ) // duplicate
    {
      arr[w++] = candidate[r];
      arr[w++] = candidate[r];
    }
    else // transpose
    {
      if ( r + 1 < candidate.length )
      {
        arr[w++] = candidate[r+1];
      }
      arr[w++] = candidate[r++];     
    }
  }
  return arr.join('');   
}

/**
 * Compute a score for the candidate string by summing 
 * the difference in encodings for each character.  
 * For example, the difference between the strings "ab" and "cd"
 * is 4, since the difference between 'a' and 'c' is 2 and
 * the difference between 'b' and 'd' is 2.  
 *
 * The lower the score, the closer the match.  A score of 0
 * indicates an exact match.
 *
 * Inputs:
 *    candidate - string to be tested
 *    target    - desired target output
 *
 * Outputs:
 *    none
 *
 * Returns: numeric score
 */
function score( candidate, target )
{
  var i = 0;
  var sum = 0;
  
  for ( i = 0; i < candidate.length && i < target.length; i++ )
  {
    sum += Math.abs( candidate.charCodeAt(i) - target.charCodeAt(i) );
  }
  while ( i < candidate.length )
  {
    sum += candidate.charCodeAt(i++);
  }
  while( i < target.length )
  {
    sum += target.charCodeAt(i++);
  }
  return sum;
}

/**
 * Initialize our weasel population by generating a bunch
 * of random strings and scoring them against the target.
 *
 * Inputs:
 *    size      - population size; this does not change over time
 *    target    - string that we are trying to match
 *    strsize   - length of initial population strings
 *    alphabet  - character set from which to generate strings
 *
 * Outputs:
 *    none
 *
 * Returns: array of weasels.
 */
function init_population(size, target, strsize, alphabet)
{
  var pop = new Array();
  var i = 0;
  for ( i = 0; i < size; i++ )
  {
    pop[i] = new Object();   
    pop[i].generation = 1;
    pop[i].string = randomString( strsize, alphabet );
    pop[i].score = score( pop[i].string, target );
  }
  
  return pop;
}
 
/**
 * Validate our input form. The target string must not contain any characters
 * that are not also in the alphabet.  The probabilities must all be between
 * 1 and 97. and they must sum to 100 or less. 
 *
 * Inputs:
 *    none
 *
 * Outputs: 
 *    none
 * 
 * Returns: true if input form passes validation, false otherwise.
 */
function validateInput()
{
  var result = true;
  
  var target = document.getElementById( "target" );
  var alphabet = document.getElementById( "alphabet" );
  var point = document.getElementById( "point" );
  var duplicate = document.getElementById( "duplicate" );
  var del = document.getElementById( "delete" );  
  var xpose = document.getElementById( "transpose" );
  var popsize = document.getElementById( "popsize" );
  var strlen = document.getElementById( "length" );
  
  for ( i = 0; i < target.value.length; i++ )
  {
    if ( alphabet.value.search( target.value[i] ) == -1 )
    {
      result = false;
      break;
    }
  }
  if ( !result )
  {
    alert( "Target contains characters that are not in the alphabet" );
  }
  var probs = point.valueAsNumber + duplicate.valueAsNumber + del.valueAsNumber + xpose.valueAsNumber;
  if ( probs > 100 )
  {
    result = false;
    alert( "Probabilities must sum to 100 or less." );
  }
  if ( point.valueAsNumber < 1 || point.valueAsNumber > 97 || 
       duplicate.valueAsNumber < 1 || duplicate.valueAsNumber > 97 || 
       del.valueAsNumber < 1 || del.valueAsNumber > 97 ||
       xpose.valueAsNumber < 1 || xpose.valueAsNumber > 97 )
  {
    result = false;
    alert( "Probability range must be between 1 and 97" );
  }
  
  if ( point.valueAsNumber > 10 || duplicate.valueAsNumber > 10 || del.valueAsNumber > 10 || xpose.valueAsNumber > 10 )
  {
    result = confirm( "Large change probability values may cause the program to run for an excessive amount of time." );
  }
  
  return result;
}
  
/**
 * Main driver, triggered by pressing OK on the input form. 
 * Validates the input form, then generates and scores the
 * initial population.  Population is sorted by score in 
 * ascending order.  
 *
 * While there is no match to the target string, discard the 90%
 * of the population with the highest scores, rebuild
 * the population from the remaining 10% by copying them
 * with mutations, score the new strings against the target,
 * and sort the population by score.   
 *
 * Each time the string with the lowest score changes, display it
 * to the output table.
 * 
 * Inputs:
 *    none
 *
 * Outputs:
 *    none
 *
 * Returns: none
 */
function doWeasel()
{
  if ( !validateInput() )
  {
    return;
  }
  
  var outputarea = document.getElementById( "outputarea" );

  var popsize = document.getElementById( "size" );
  var target = document.getElementById( "target" );
  var alphabet = document.getElementById( "alphabet" );
  var strsize = document.getElementById( "length" );
  
  var table = document.getElementById( "outputTable" );
  if ( table != undefined )
  {
    outputarea.removeChild( table );
  }
  
  var prob_table = [96, 97, 98, 99];
  var param = document.getElementById ( "transpose" );
  prob_table[3] = 100 - param.value;
  param = document.getElementById( "delete" );
  prob_table[2] = prob_table[3] - param.value;
  param = document.getElementById( "duplicate" );
  prob_table[1] = prob_table[2] - param.value;
  param = document.getElementById ("point" );
  prob_table[0] = prob_table[1] - param.value;
  
  
  var population = init_population( popsize.value, target.value, strsize.value, alphabet.value );
  population.sort( function(a,b) { return a.score - b.score; });
  
  var table = document.createElement( "table" );
  table.setAttribute( "id", "outputTable" );
  var head = table.createTHead();
  var row = head.insertRow();
  var cell = row.insertCell();
  cell.innerHTML = "Generation";
  cell = row.insertCell();
  cell.innerHTML = "Score";
  cell = row.insertCell();
  cell.innerHTML = "String";
  table.appendChild( head );
  
  row = table.insertRow();
  
  cell = row.insertCell();
  cell.innerHTML = population[0].generation;
  cell = row.insertCell();
  cell.innerHTML = population[0].score;
  cell = row.insertCell();
  cell.innerHTML = population[0].string;
  var p;
  var g = 1;
  var breeders = Math.floor( population.length / 10 );
  
  var best_fit = population[0].string;
  
  while( population[0].score > 0 )
  {
    g++;
    for ( p = breeders; p < population.length; p++ )
    {
      population[p].generation = g;
      population[p].string = mutate( population[p%breeders].string, prob_table, alphabet.value );
      population[p].score = score( target.value, population[p].string );
    }
    population.sort( function(a,b) { return a.score - b.score; } );
    
    if ( population[0].string != best_fit )
    {
      row = table.insertRow();
      cell = row.insertCell();
      cell.innerHTML = population[0].generation;
      cell = row.insertCell();
      cell.innerHTML = population[0].score;
      cell = row.insertCell();
      cell.innerHTML = population[0].string;
      best_fit = population[0].string;
    }
  }  
  outputarea.append( table );
}
